imap_processing/hi/l1b/hi_l1b.py

```diff
@@ -90,28 +107,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
     l1b_dataset : xarray.Dataset
         L1B direct event data.
     """
-    n_epoch = l1a_dataset["epoch"].size
+    l1b_dataset = compute_coincidence_type_and_time_deltas(l1a_dataset)
     l1b_de_var_names = [
-        "coincidence_type",
         "esa_energy_step",
-        "delta_t_ab",
-        "delta_t_ac1",
-        "delta_t_bc1",
-        "delta_t_c1c2",
         "spin_phase",
         "hae_latitude",
         "hae_longitude",
         "quality_flag",
         "nominal_bin",
     ]
     new_data_vars = create_dataset_variables(
-        l1b_de_var_names, (n_epoch,), att_manager_lookup_str="hi_de_{0}"
+        l1b_de_var_names, l1a_dataset["epoch"].size, att_manager_lookup_str="hi_de_{0}"
     )
-    l1b_dataset = l1a_dataset.assign(new_data_vars)
+    l1b_dataset = l1b_dataset.assign(new_data_vars)
     l1b_dataset = l1b_dataset.drop_vars(
         ["tof_1", "tof_2", "tof_3", "de_tag", "ccsds_met", "meta_event_met"]
     )
 
     de_global_attrs = ATTR_MGR.get_global_attributes("imap_hi_l1b_de_attrs")
     l1b_dataset.attrs.update(**de_global_attrs)
     return l1b_dataset
+
+
+def compute_coincidence_type_and_time_deltas(dataset: xr.Dataset) -> xr.Dataset:
+    """
+    Compute coincidence type and time deltas.
+
+    Adds the new variables "coincidence_type", "delta_t_ab", "delta_t_ac1",
+    "delta_t_bc1", and "delta_t_c1c2" to the input xarray.Dataset and returns
+    the updated xarray.Dataset.
+
+    Parameters
+    ----------
+    dataset : xarray.Dataset
+        The L1A/B dataset that results from reading in the L1A CDF and
+        allocating the new L1B DataArrays.
+
+    Returns
+    -------
+    xr.Dataset
+        Updated xarray.Dataset with 5 new variables added.
+    """
+    new_data_vars = create_dataset_variables(
+        [
+            "coincidence_type",
+            "delta_t_ab",
+            "delta_t_ac1",
+            "delta_t_bc1",
+            "delta_t_c1c2",
+        ],
+        len(dataset.epoch),
+        "hi_de_{0}",
+    )
+    out_ds = dataset.assign(new_data_vars)
+
+    # compute masks needed for coincidence type and delta t calculations
+    a_mask = out_ds.trigger_id.values == TriggerId.A
+    b_mask = out_ds.trigger_id.values == TriggerId.B
+    c_mask = out_ds.trigger_id.values == TriggerId.C
+
+    tof_1_valid_mask = np.isin(
+        out_ds.tof_1.values, HiConstants.TOF1_BAD_VALUES, invert=True
+    )
+    tof_2_valid_mask = np.isin(
+        out_ds.tof_2.values, HiConstants.TOF2_BAD_VALUES, invert=True
+    )
+    tof_1and2_valid_mask = tof_1_valid_mask & tof_2_valid_mask
+    tof_3_valid_mask = np.isin(
+        out_ds.tof_3.values, HiConstants.TOF3_BAD_VALUES, invert=True
+    )
+
+    # Table denoting how hit-first mask and valid TOF masks are used to set
+    # coincidence type bitmask
+    # -----------------------------------------------------------------------
```

Comment:
I like this table, but I'm a little confused how this relates to the code below. For line 189 it looks like the coincidence type is being set  to the A value for all indices where either `a_mask` or `tof_1_valid_mask` are true. But I'm not sure how to get that from the table.

Reply:
I wondered if that implementation would cause confusion. The union of events where `a_mask == True` with events where `a_mask == False & tof_1_valid_mask == True` is the same as the union of events where `a_mask == True` with events where `tof_1_valid_mask == True` because the events in the `a_mask == True` set contain both `a_mask == True & tof_1_valid_mask == True` and `a_mask == True & tof_1_valid_mask == False`. I thought I would be clever and make the code more succinct, but perhaps it just makes it more confusing?

Reply:
Hopefully, adding both coincidence type bits implied by valid TOF values to the table made it easier to see? It made me catch a simplification to the C1 mask.